pkg/engine/engine.go

```diff
@@ -1,26 +1,94 @@
 package engine
 
-import "github.com/grafana/loki/v3/pkg/logql/syntax"
-
-// canExecuteWithNewEngine determines whether a query can be executed by the new execution engine.
-func canExecuteWithNewEngine(expr syntax.Expr) bool {
-	switch expr := expr.(type) {
-	case syntax.SampleExpr:
-		return false
-	case syntax.LogSelectorExpr:
-		ret := true
-		expr.Walk(func(e syntax.Expr) bool {
-			switch e.(type) {
-			case *syntax.LineParserExpr, *syntax.LogfmtParserExpr, *syntax.LogfmtExpressionParserExpr, *syntax.JSONExpressionParserExpr:
-				ret = false
-			case *syntax.LineFmtExpr, *syntax.LabelFmtExpr:
-				ret = false
-			case *syntax.KeepLabelsExpr, *syntax.DropLabelsExpr:
-				ret = false
-			}
-			return true
-		})
-		return ret
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/metastore"
+	"github.com/grafana/loki/v3/pkg/engine/planner/logical"
+	"github.com/grafana/loki/v3/pkg/engine/planner/physical"
+	"github.com/grafana/loki/v3/pkg/logql"
+	"github.com/grafana/loki/v3/pkg/logqlmodel"
+	utillog "github.com/grafana/loki/v3/pkg/util/log"
+)
+
+var (
+	ErrNotSupported = errors.New("feature not supported in new query engine")
+)
+
+// New creates a new instance of the query engine that implements the [logql.Engine] interface.
+func New(opts logql.EngineOpts, metastore metastore.Metastore, limits logql.Limits, logger log.Logger) *QueryEngine {
+	return &QueryEngine{
+		logger:    logger,
+		limits:    limits,
+		metastore: metastore,
+		opts:      opts,
+	}
+}
+
+// QueryEngine combines logical planning, physical planning, and execution to evaluate LogQL queries.
+type QueryEngine struct {
+	logger    log.Logger
+	limits    logql.Limits
+	metastore metastore.Metastore
+	opts      logql.EngineOpts
+}
+
+// Query implements [logql.Engine].
+func (e *QueryEngine) Query(params logql.Params) logql.Query {
+	return &queryAdapter{
+		engine: e,
+		params: params,
 	}
-	return false
 }
+
+// Execute executes a LogQL query and returns its results or alternatively an error.
+// The execution is done in three steps:
+//  1. Create a logical plan from the provided query parameters.
+//  2. Create a physical plan from the logical plan using information from the catalog.
+//  3. Evaluate the physical plan with the executor.
+func (e *QueryEngine) Execute(ctx context.Context, params logql.Params) (logqlmodel.Result, error) {
```

Comment:
nit: Do we want metrics for latency + how many queries were supported/unsupported? (It doesn't have to be in this PR if you think it should wait)

Reply:
We should have these metrics, yes. Let's add them in a separate PR.